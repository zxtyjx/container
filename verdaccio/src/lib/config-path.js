// @flow

import fs from 'fs';
import _ from 'lodash';
import Path from 'path';
import logger from './logger';
import mkdirp from 'mkdirp';

import {folder_exists, fileExists} from './utils';

const CONFIG_FILE = 'config.yaml';
const XDG = 'xdg';
const WIN = 'win';
const WIN32 = 'win32';
const pkgJSON = require('../../package.json');

/**
 * Find and get the first config file that match.
 * @return {String} the config file path
 */
function findConfigFile(configPath: any) {
  if (_.isNil(configPath) === false) {
    return Path.resolve(configPath);
  }

  const configPaths = getConfigPaths();

  if (_.isEmpty(configPaths)) {
    throw new Error('no configuration files can be proccesed');
  }

  const primaryConf: any = _.find(configPaths, (configLocation: any) => fileExists(configLocation.path));
  if (_.isNil(primaryConf) === false) {
    return primaryConf.path;
  }

  return createConfigFile(_.head(configPaths)).path;
}

function createConfigFile(configLocation: any) {
  createConfigFolder(configLocation);

  const defaultConfig = updateStorageLinks(configLocation, readDefaultConfig());

  fs.writeFileSync(configLocation.path, defaultConfig);

  return configLocation;
}

function readDefaultConfig() {
  return fs.readFileSync(require.resolve('../../conf/default.yaml'), 'utf8');
}

function createConfigFolder(configLocation) {
  mkdirp.sync(Path.dirname(configLocation.path));
  logger.logger.info({file: configLocation.path}, 'Creating default config file in @{file}');
}

function updateStorageLinks(configLocation, defaultConfig) {
    if (configLocation.type !== XDG) {
        return defaultConfig;
    }

    // $XDG_DATA_HOME defines the base directory relative to which user specific data files should be stored,
    // If $XDG_DATA_HOME is either not set or empty, a default equal to $HOME/.local/share should be used.
    // $FlowFixMe
    let dataDir = process.env.XDG_DATA_HOME || Path.join(process.env.HOME, '.local', 'share');
    if (folder_exists(dataDir)) {
      dataDir = Path.resolve(Path.join(dataDir, pkgJSON.name, 'storage'));
      return defaultConfig.replace(/^storage: .\/storage$/m, `storage: ${dataDir}`);
    } else {
      return defaultConfig;
    }
}

function getConfigPaths() {
  return [getXDGDirectory(), getWindowsDirectory(), getRelativeDefaultDirectory(), getOldDirectory()].filter((path) => !!path);
}

const getXDGDirectory = () => {
  const XDGConfig = getXDGHome() ||
        process.env.HOME && Path.join(process.env.HOME, '.config');

  if (XDGConfig && folder_exists(XDGConfig)) {
    return {
      path: Path.join(XDGConfig, pkgJSON.name, CONFIG_FILE),
      type: XDG,
    };
  }
};

const getXDGHome = () => process.env.XDG_CONFIG_HOME;

const getWindowsDirectory = () => {
  if (process.platform === WIN32 && process.env.APPDATA && folder_exists(process.env.APPDATA)) {
    return {
      path: Path.resolve(Path.join(process.env.APPDATA, pkgJSON.name, CONFIG_FILE)),
      type: WIN,
    };
  }
};

const getRelativeDefaultDirectory = () => {
  return {
    path: Path.resolve(Path.join('.', pkgJSON.name, CONFIG_FILE)),
    type: 'def',
  };
};

const getOldDirectory = () => {
  return {
    path: Path.resolve(Path.join('.', CONFIG_FILE)),
    type: 'old',
  };
};

export default findConfigFile;
