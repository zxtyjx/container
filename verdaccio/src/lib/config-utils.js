// @flow
import _ from 'lodash';
import assert from 'assert';
import minimatch from 'minimatch';

import {ErrorCode} from './utils';

import type {PackageList, UpLinksConfList} from '@verdaccio/types';
import type {MatchedPackage} from '../../types';

const BLACKLIST = {
  all: true,
  anonymous: true,
  undefined: true,
  owner: true,
  none: true,
};

/**
 * Normalise user list.
 * @return {Array}
 */
export function normalizeUserlist(oldFormat: any, newFormat: any) {
  const result = [];
  /* eslint prefer-rest-params: "off" */

  for (let i=0; i < arguments.length; i++) {
    if (arguments[i] == null) {
      continue;
    }

    // if it's a string, split it to array
    if (_.isString(arguments[i])) {
      result.push(arguments[i].split(/\s+/));
    } else if (Array.isArray(arguments[i])) {
      result.push(arguments[i]);
    } else {
      throw ErrorCode.getInternalError('CONFIG: bad package acl (array or string expected): ' + JSON.stringify(arguments[i]));
    }
  }
  return _.flatten(result);
}

export function uplinkSanityCheck(uplinks: UpLinksConfList, users: any = BLACKLIST) {
  const newUplinks = _.clone(uplinks);
  let newUsers = _.clone(users);

  for (let uplink in newUplinks) {
    if (Object.prototype.hasOwnProperty.call(newUplinks, uplink)) {
      if (_.isNil(newUplinks[uplink].cache)) {
        newUplinks[uplink].cache = true;
      }
      newUsers = sanityCheckNames(uplink, newUsers);
    }
  }

  return newUplinks;
}

export function sanityCheckNames(item: string, users: any) {
  assert(item !== 'all' && item !== 'owner' && item !== 'anonymous' && item !== 'undefined' && item !== 'none', 'CONFIG: reserved uplink name: ' + item);
  assert(!item.match(/\s/), 'CONFIG: invalid uplink name: ' + item);
  assert(_.isNil(users[item]), 'CONFIG: duplicate uplink name: ' + item);
  users[item] = true;

  return users;
}

export function sanityCheckUplinksProps(configUpLinks: any) {
  const uplinks = _.clone(configUpLinks);

  for (let uplink in uplinks) {
    if (Object.prototype.hasOwnProperty.call(uplinks, uplink)) {
      assert(uplinks[uplink].url, 'CONFIG: no url for uplink: ' + uplink);
      assert( _.isString(uplinks[uplink].url), 'CONFIG: wrong url format for uplink: ' + uplink);
      uplinks[uplink].url = uplinks[uplink].url.replace(/\/$/, '');
    }
  }

  return uplinks;
}

/**
 * Check whether an uplink can proxy
 */
export function hasProxyTo(pkg: string, upLink: string, packages: PackageList): boolean {
  const matchedPkg: MatchedPackage = (getMatchedPackagesSpec(pkg, packages): MatchedPackage);
  const proxyList = typeof matchedPkg !== 'undefined' ? matchedPkg.proxy : [];
  if (proxyList) {
    return proxyList.some((curr) => upLink === curr);
  }

  return false;
}

export function getMatchedPackagesSpec(pkg: string, packages: PackageList): MatchedPackage {
  for (let i in packages) {
    // $FlowFixMe
    if (minimatch.makeRe(i).exec(pkg)) {
      return packages[i];
    }
  }
  return;
}

export function normalisePackageAccess(packages: PackageList): PackageList {
  const normalizedPkgs: PackageList = {...packages};
  // add a default rule for all packages to make writing plugins easier
  if (_.isNil(normalizedPkgs['**'])) {
    normalizedPkgs['**'] = {};
  }

  for (let pkg in packages) {
    if (Object.prototype.hasOwnProperty.call(packages, pkg)) {
      assert(_.isObject(packages[pkg]) && _.isArray(packages[pkg]) === false,
        `CONFIG: bad "'${pkg}'" package description (object expected)`);
      normalizedPkgs[pkg].access = normalizeUserlist(packages[pkg].allow_access, packages[pkg].access);
      delete normalizedPkgs[pkg].allow_access;
      normalizedPkgs[pkg].publish = normalizeUserlist(packages[pkg].allow_publish, packages[pkg].publish);
      delete normalizedPkgs[pkg].allow_publish;
      normalizedPkgs[pkg].proxy = normalizeUserlist(packages[pkg].proxy_access, packages[pkg].proxy);
      delete normalizedPkgs[pkg].proxy_access;
    }
  }

  return normalizedPkgs;
}

