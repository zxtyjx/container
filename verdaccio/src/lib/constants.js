// @flow

export const DEFAULT_PORT = '4873';
export const DEFAULT_DOMAIN = 'localhost';

export const HEADERS = {
  JSON: 'application/json',
  JSON_CHARSET: 'application/json; charset=utf-8',
  OCTET_STREAM: 'application/octet-stream; charset=utf-8',
  TEXT_CHARSET: 'text/plain; charset=utf-8',
  GZIP: 'gzip',
};

export const HEADER_TYPE = {
  CONTENT_ENCODING: 'content-encoding',
  CONTENT_TYPE: 'content-type',
  CONTENT_LENGTH: 'content-length',
  ACCEPT_ENCODING: 'accept-encoding',
};

export const ERROR_CODE = {
  token_required: 'token is required',
};

export const TOKEN_BASIC = 'Basic';
export const TOKEN_BEARER = 'Bearer';
export const DEFAULT_REGISTRY = 'https://registry.npmjs.org';
export const DEFAULT_UPLINK = 'npmjs';

export const ROLES = {
  $ALL: '$all',
  ALL: 'all',
  $AUTH: '$authenticated',
  $ANONYMOUS: '$anonymous',
  DEPRECATED_ALL: '@all',
  DEPRECATED_AUTH: '@authenticated',
  DEPRECATED_ANONUMOUS: '@anonymous',
};

export const HTTP_STATUS = {
  OK: 200,
  CREATED: 201,
  MULTIPLE_CHOICES: 300,
  NOT_MODIFIED: 304,
  BAD_REQUEST: 400,
  UNAUTHORIZED: 401,
  FORBIDDEN: 403,
  NOT_FOUND: 404,
  CONFLICT: 409,
  UNSUPORTED_MEDIA: 415,
  BAD_DATA: 422,
  INTERNAL_ERROR: 500,
  SERVICE_UNAVAILABLE: 503,
  LOOP_DETECTED: 508,
};

export const API_MESSAGE = {
  PKG_CREATED: 'created new package',
  PKG_CHANGED: 'package changed',
  PKG_REMOVED: 'package removed',
  PKG_PUBLISHED: 'package published',
  TARBALL_REMOVED: 'tarball removed',
  TAG_UPDATED: 'tags updated',
  TAG_REMOVED: 'tag removed',
  TAG_ADDED: 'package tagged',

};

export const API_ERROR = {
  BAD_USERNAME_PASSWORD: 'bad username/password, access denied',
  NO_PACKAGE: 'no such package available',
  NOT_ALLOWED: 'not allowed to access package',
  INTERNAL_SERVER_ERROR: 'internal server error',
  UNKNOWN_ERROR: 'unknown error',
  NOT_PACKAGE_UPLINK: 'package does not exist on uplink',
  UPLINK_OFFLINE_PUBLISH: 'one of the uplinks is down, refuse to publish',
  UPLINK_OFFLINE: 'uplink is offline',
  CONTENT_MISMATCH: 'content length mismatch',
  NOT_FILE_UPLINK: 'file doesn\'t exist on uplink',
  MAX_USERS_REACHED: 'maximum amount of users reached',
  VERSION_NOT_EXIST: 'this version doesn\'t exist',
  FILE_NOT_FOUND: 'File not found',
  BAD_STATUS_CODE: 'bad status code',
  PACKAGE_EXIST: 'this package is already present',
  BAD_AUTH_HEADER: 'bad authorization header',
  WEB_DISABLED: 'Web interface is disabled in the config file',
  DEPRECATED_BASIC_HEADER: 'basic authentication is deprecated, please use JWT instead',
  BAD_FORMAT_USER_GROUP: 'user groups is different than an array',
};

export const APP_ERROR = {
  CONFIG_NOT_VALID: 'CONFIG: it does not look like a valid config file',
};

export const DEFAULT_NO_README = 'ERROR: No README data found!';
export const MODULE_NOT_FOUND = 'MODULE_NOT_FOUND';

export const WEB_TITLE = 'Verdaccio';

export const PACKAGE_ACCESS = {
  SCOPE: '@*/*',
  ALL: '**',
};
