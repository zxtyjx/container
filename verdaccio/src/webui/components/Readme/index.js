
import React from 'react';
import PropTypes from 'prop-types';

import 'github-markdown-css';

const Readme = (props) => {
  return <div className="markdown-body" dangerouslySetInnerHTML={{__html: props.readMe}}/>;
};

Readme.propTypes = {
  readMe: PropTypes.string.isRequired
};

export default Readme;
