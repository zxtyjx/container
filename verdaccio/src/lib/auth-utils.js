// @flow

import {ErrorCode} from './utils';
import {API_ERROR} from './constants';

import type {RemoteUser, Package, Callback} from '@verdaccio/types';

export function allow_action(action: string) {
  return function(user: RemoteUser, pkg: Package, callback: Callback) {
    const {name, groups} = user;
    const hasPermission = pkg[action].some((group) => name === group || groups.includes(group));

    if (hasPermission) {
      return callback(null, true);
    }

    if (name) {
      callback(ErrorCode.getForbidden(`user ${name} is not allowed to ${action} package ${pkg.name}`));
    } else {
      callback(ErrorCode.getForbidden(`unregistered users are not allowed to ${action} package ${pkg.name}`));
    }
  };
}

export function getDefaultPlugins() {
  return {
    authenticate(user: string, password: string, cb: Callback) {
      cb(ErrorCode.getForbidden(API_ERROR.BAD_USERNAME_PASSWORD));
    },

    add_user(user: string, password: string, cb: Callback) {
      return cb(ErrorCode.getConflict(API_ERROR.BAD_USERNAME_PASSWORD));
    },

    allow_access: allow_action('access'),
    allow_publish: allow_action('publish'),
  };
}
