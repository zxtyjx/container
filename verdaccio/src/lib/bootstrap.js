// @flow

import {assign, isObject, isFunction} from 'lodash';
import Path from 'path';
import URL from 'url';
import fs from 'fs';
import http from 'http';
import https from 'https';
// $FlowFixMe
import constants from 'constants';
import endPointAPI from '../api/index';
import {parse_address} from './utils';

import type {Callback} from '@verdaccio/types';
import type {$Application} from 'express';
import {DEFAULT_PORT} from './constants';

const logger = require('./logger');

/**
 * Retrieve all addresses defined in the config file.
 * Verdaccio is able to listen multiple ports
 * @param {String} argListen
 * @param {String} configListen
 * eg:
 *  listen:
    - localhost:5555
    - localhost:5557
    @return {Array}
 */
export function getListListenAddresses(argListen: string, configListen: mixed) {
  // command line || config file || default
  let addresses;
  if (argListen) {
    addresses = [argListen];
  } else if (Array.isArray(configListen)) {
    addresses = configListen;
  } else if (configListen) {
    addresses = [configListen];
  } else {
    addresses = [DEFAULT_PORT];
  }
  addresses = addresses.map(function(addr) {
    const parsedAddr = parse_address(addr);

    if (!parsedAddr) {
      logger.logger.warn({addr: addr},
         'invalid address - @{addr}, we expect a port (e.g. "4873"),'
       + ' host:port (e.g. "localhost:4873") or full url'
       + ' (e.g. "http://localhost:4873/")');
    }

    return parsedAddr;
  }).filter(Boolean);

  return addresses;
}

/**
 * Trigger the server after configuration has been loaded.
 * @param {Object} config
 * @param {Object} cliArguments
 * @param {String} configPath
 * @param {String} pkgVersion
 * @param {String} pkgName
 */
function startVerdaccio(config: any,
                        cliListen: string,
                        configPath: string,
                        pkgVersion: string,
                        pkgName: string,
                        callback: Callback) {
  if (isObject(config) === false) {
    throw new Error('config file must be an object');
  }

  endPointAPI(config).then((app)=> {
    const addresses = getListListenAddresses(cliListen, config.listen);

    addresses.forEach(function(addr) {
      let webServer;
      if (addr.proto === 'https') {
        // https  must either have key cert and ca  or a pfx and (optionally) a passphrase
        if (!config.https || !config.https.key || !config.https.cert || !config.https.ca) {
          displayHTTPSWarning(configPath);
        }

        webServer = handleHTTPS(app, configPath, config);
      } else { // http
        webServer = http.createServer(app);
      }

      unlinkAddressPath(addr);

      callback(webServer, addr, pkgName, pkgVersion);
    });
  });
}

function unlinkAddressPath(addr) {
  if (addr.path && fs.existsSync(addr.path)) {
    fs.unlinkSync(addr.path);
  }
}

function displayHTTPSWarning(storageLocation) {
  const resolveConfigPath = function(file) {
    return Path.resolve(Path.dirname(storageLocation), file);
  };

  logger.logger.fatal([
    'You have enabled HTTPS and need to specify either ',
    '    "https.key", "https.cert" and "https.ca" or ',
    '    "https.pfx" and optionally "https.passphrase" ',
    'to run https server',
    '',
    // commands are borrowed from node.js docs
    'To quickly create self-signed certificate, use:',
    ' $ openssl genrsa -out ' + resolveConfigPath('verdaccio-key.pem') + ' 2048',
    ' $ openssl req -new -sha256 -key ' + resolveConfigPath('verdaccio-key.pem') + ' -out ' + resolveConfigPath('verdaccio-csr.pem'),
    ' $ openssl x509 -req -in ' + resolveConfigPath('verdaccio-csr.pem') +
    ' -signkey ' + resolveConfigPath('verdaccio-key.pem') + ' -out ' + resolveConfigPath('verdaccio-cert.pem'),
    '',
    'And then add to config file (' + storageLocation + '):',
    '  https:',
    `    key: ${resolveConfigPath('verdaccio-key.pem')}`,
    `    cert: ${resolveConfigPath('verdaccio-cert.pem')}`,
    `    ca: ${resolveConfigPath('verdaccio-csr.pem')}`,
  ].join('\n'));
  process.exit(2);
}

function handleHTTPS(app, configPath, config) {
  try {
    let httpsOptions = {
      secureProtocol: 'SSLv23_method', // disable insecure SSLv2 and SSLv3
      secureOptions: constants.SSL_OP_NO_SSLv2 | constants.SSL_OP_NO_SSLv3,
    };

    if (config.https.pfx) {
      httpsOptions = assign(httpsOptions, {
        pfx: fs.readFileSync(config.https.pfx),
        passphrase: config.https.passphrase || '',
      });
    } else {
      httpsOptions = assign(httpsOptions, {
        key: fs.readFileSync(config.https.key),
        cert: fs.readFileSync(config.https.cert),
        ca: fs.readFileSync(config.https.ca),
      });
    }
    return https.createServer(httpsOptions, app);
  } catch (err) { // catch errors related to certificate loading
    logger.logger.fatal({err: err}, 'cannot create server: @{err.message}');
    process.exit(2);
  }
}

function listenDefaultCallback(webServer: $Application, addr: any, pkgName: string, pkgVersion: string) {
  webServer.listen(addr.port || addr.path, addr.host, () => {
    // send a message for tests
    if (isFunction(process.send)) {
      process.send({
        verdaccio_started: true,
      });
    }
  // $FlowFixMe
  }).on('error', function(err) {
    logger.logger.fatal({err: err}, 'cannot create server: @{err.message}');
    process.exit(2);
  });

  logger.logger.warn({
    addr: ( addr.path
          ? URL.format({
              protocol: 'unix',
              pathname: addr.path,
            })
          : URL.format({
              protocol: addr.proto,
              hostname: addr.host,
              port: addr.port,
              pathname: '/',
            })
          ),
    version: pkgName + '/' + pkgVersion,
  }, 'http address - @{addr} - @{version}');
}

export {startVerdaccio, listenDefaultCallback};
