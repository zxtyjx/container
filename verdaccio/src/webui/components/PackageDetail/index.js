import React from 'react';
import PropTypes from 'prop-types';
import isNil from 'lodash/isNil';

import Readme from '../Readme';

import classes from './packageDetail.scss';

const displayState = (readMe) => {
  return !isNil(readMe) ? <Readme readMe={readMe} /> : '';
};

const PackageDetail = ({packageName, readMe}) => {
  return (
    <div className={classes.pkgDetail}>
      <h1 className={classes.title}>{packageName}</h1>
      <div className={classes.readme}>{displayState(readMe)}</div>
    </div>
  );
};

PackageDetail.propTypes = {
  readMe: PropTypes.string,
  packageName: PropTypes.string.isRequired
};

export default PackageDetail;
