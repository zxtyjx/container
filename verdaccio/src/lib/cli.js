#!/usr/bin/env node

/* eslint no-sync:0 */
/* eslint no-empty:0 */

import path from 'path';
import {startVerdaccio, listenDefaultCallback} from './bootstrap';
import findConfigFile from './config-path';

if (process.getuid && process.getuid() === 0) {
  global.console.error('Verdaccio doesn\'t need superuser privileges. Don\'t run it under root.');
}

process.title = 'verdaccio';

try {
  // for debugging memory leaks
  // totally optional
  require('heapdump');
} catch (err) { }

const logger = require('./logger');
logger.setup(); // default setup

const commander = require('commander');
const Utils = require('./utils');
const pkginfo = require('pkginfo')(module); // eslint-disable-line no-unused-vars
const pkgVersion = module.exports.version;
const pkgName = module.exports.name;

commander
  .option('-l, --listen <[host:]port>', 'host:port number to listen on (default: localhost:4873)')
  .option('-c, --config <config.yaml>', 'use this configuration file (default: ./config.yaml)')
  .version(pkgVersion)
  .parse(process.argv);

if (commander.args.length == 1 && !commander.config) {
  // handling "verdaccio [config]" case if "-c" is missing in commandline
  commander.config = commander.args.pop();
}

if (commander.args.length !== 0) {
  commander.help();
}
let verdaccioConfiguration;
let configPathLocation;
const cliListner = commander.listen;

try {
  configPathLocation = findConfigFile(commander.config);
  verdaccioConfiguration = Utils.parseConfigFile(configPathLocation);
  process.title = verdaccioConfiguration.web && verdaccioConfiguration.web.title || 'verdaccio';

  if (!verdaccioConfiguration.self_path) {
    verdaccioConfiguration.self_path = path.resolve(configPathLocation);
  }
  if (!verdaccioConfiguration.https) {
    verdaccioConfiguration.https = {enable: false};
  }

  logger.logger.warn({file: configPathLocation}, 'config file  - @{file}');

  startVerdaccio(verdaccioConfiguration, cliListner, configPathLocation, pkgVersion, pkgName, listenDefaultCallback);
} catch (err) {
  logger.logger.fatal({file: configPathLocation, err: err}, 'cannot open config file @{file}: @{!err.message}');
  process.exit(1);
}

process.on('uncaughtException', function(err) {
  logger.logger.fatal( {
      err: err,
    },
  'uncaught exception, please report this\n@{err.stack}' );
  process.exit(255);
});
