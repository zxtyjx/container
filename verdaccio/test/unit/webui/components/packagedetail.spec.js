/**
 * PackageDetail component
 */
import React from 'react';
import { shallow } from 'enzyme';
import PackageDetail from '../../../../src/webui/components/PackageDetail/index';
import Readme from '../../../../src/webui/components/Readme/index';
import {WEB_TITLE} from '../../../../src/lib/constants';

console.error = jest.fn();

describe('<PackageDetail /> component', () => {
  it('should give error for required props', () => {
    shallow(<PackageDetail />);
    expect(console.error).toBeCalled();
  });

  it('should load the component', () => {
    const props = {
      readMe: 'Test readme',
      packageName: WEB_TITLE
    };
    const wrapper = shallow(<PackageDetail {...props} />);

    expect(wrapper.find('h1').text()).toEqual(WEB_TITLE);
    expect(
      wrapper
        .find(Readme)
        .dive()
        .html()
    ).toEqual('<div class="markdown-body">Test readme</div>');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
