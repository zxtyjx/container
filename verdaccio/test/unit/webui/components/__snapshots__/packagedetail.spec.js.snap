// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PackageDetail /> component should load the component 1`] = `"<div class=\\"pkgDetail\\"><h1 class=\\"title\\">Verdaccio</h1><div class=\\"readme\\"><div class=\\"markdown-body\\">Test readme</div></div></div>"`;
